File: examples/grammar/fractional-range-values/python/_cott_impl/curriculum/fractional_range_values/build_bounded_range.py
from cott_runtime import CottList, Err, F64, Ok, Result
from curriculum.fractional_range_values_types import FractionalRangeError, FractionalRangeError_NonFiniteInput, FractionalRangeError_OutputLimitExceeded, FractionalRangeError_StepDoesNotAdvance, MAX_F64, OutputLimit, PositiveStep


def build_bounded_range(start: F64, stop: F64, step: PositiveStep, limit: OutputLimit) -> Result[CottList[F64], FractionalRangeError]:
    step_value = step.value
    if not (-MAX_F64 <= start <= MAX_F64) or not (-MAX_F64 <= stop <= MAX_F64) or step_value > MAX_F64:
        return Err(error=FractionalRangeError_NonFiniteInput())

    values: list[F64] = []
    count = limit.value
    index = 0
    while True:
        product = float(index) * step_value
        candidate = start + product
        if not candidate < stop:
            return Ok(value=CottList(values=values))
        if index == count:
            return Err(error=FractionalRangeError_OutputLimitExceeded())
        if values and candidate <= values[-1]:
            return Err(error=FractionalRangeError_StepDoesNotAdvance())
        values.append(candidate)
        index += 1

File: examples/kotlin/grammar/fractional-range-values/kotlin/cott_impl/curriculum/fractional_range_values/build_bounded_range.kt
package cott_impl.curriculum.fractional_range_values

internal fun build_bounded_range(start: kotlin.Double, stop: kotlin.Double, step: curriculum.fractional_range_values.PositiveStep, limit: curriculum.fractional_range_values.OutputLimit): cott_runtime.CottResult<cott_runtime.CottList<kotlin.Double>, curriculum.fractional_range_values.FractionalRangeError> {
    val stepValue: kotlin.Double = step.value
    if (!start.isFinite() || !stop.isFinite() || !stepValue.isFinite()) {
        return cott_runtime.Err(curriculum.fractional_range_values.FractionalRangeError.NonFiniteInput)
    }
    val values = kotlin.collections.ArrayList<kotlin.Double>()
    var index: kotlin.ULong = 0uL
    var previous: kotlin.Double = start
    while (true) {
        val product: kotlin.Double = index.toDouble() * stepValue
        val candidate: kotlin.Double = start + product
        if (!(candidate < stop)) {
            return cott_runtime.Ok(cott_runtime.CottList(values))
        }
        if (index == limit.value) {
            return cott_runtime.Err(curriculum.fractional_range_values.FractionalRangeError.OutputLimitExceeded)
        }
        if (index > 0uL && candidate <= previous) {
            return cott_runtime.Err(curriculum.fractional_range_values.FractionalRangeError.StepDoesNotAdvance)
        }
        values.add(candidate)
        previous = candidate
        index += 1uL
    }
}
